### DISCUSSION TITLE
Is it possible to use this library in a hybrid environment?

### DISCUSSION THREAD

**Discussion by mkrotel14** [jpudysz/react-native-unistyles#195] [Q&A] [answered]
I work in a company that uses both Android and iOS as main mobile technology, eventually some features are added in react native to speed up the development process.

My question then is, Can I use this library in a react native that was manually integrated into Android/iOS? I tried in a bare react-native project and it works but I can't seem to get past the "Unistyles runtime is not available..." error.

My React Native projects starts with each "context" being isolated one from the other

```js
// index.js

import {AppRegistry} from 'react-native';

import App from './src/App';
import Feature1 from './src/pages/Feature1';
import Feature2 from './src/pages/Feature2';

function FeatureRegister1(props) {
    return (
        <App {...props}>
            <Feature1 />
        </App>
    )
}

function FeatureRegister2(props) {
    return (
        <App {...props}>
            <Feature2 />
        </App>
    )
}

AppRegistry.registerComponent("Feature1", () => FeatureRegister1)
AppRegistry.registerComponent("Feature2", () => FeatureRegister2)
```

This enables for `Feature1` and `Feature2` have its own routes and screen without sharing context or having props that are unnecessary (although the bundle is the same, we are not doing code splitting for this example).

The App is a basic wrapper for common code between `Feature1` and `Feature2` where I initiate useQuery, react-intl and the Unistyle.

I don't know how can I make it work but it seems that all efforts lead to nowhere.

My folder structure looks like:
```
  - src/
    - App.tsx
    - pages/
    - shared/
      - theme
        - index.ts
  package.json
  unistyles.ts
```

My unistyle file imports themes defined inside the `theme.ts`
```ts
// unistyle.ts
import {UnistylesRegistry} from 'react-native-unistyles';
import {theme} from './src/shared/theme';

type AppThemes = {
  customTheme: typeof theme;
};

declare module 'react-native-unistyles' {
  export interface UnistylesThemes extends AppThemes {}
}

UnistylesRegistry.addThemes({
  customTheme: theme,
}).addConfig({
  initialTheme: 'customTheme',
});
```

Not to mention that my `App.tsx` contains the unistyle import

```tsx
// App.tsx
import '../unistlyles';

function App({children}): JSX.Element {
  return (
    <View>
      {children}
    </View>
  );
}

export default App;
```

**Comment by jpudysz** [comment]
Hey @mkrotel14  to be honest, I don't know. Your setup seems to be correct, but the "Unistyles runtime is not available..." error is present when your iOS/Android projects didn't detect Unistyles via autolinking. This may happen if your hybrid app somehow doesn't support autolinking.

Here are the steps required to give you some insight:
- Install Unistyles and ensure it's listed as a dependency in package.json.
- Run `pod install` to install the native code; since Unistyles contains a `podspec` file, it will be auto-linked by React Native.
- During the` pod install`, you should see a message stating that `react-native-unistyles (2.7.1)` has been installed.
- At this point, the "Unistyles runtime is not available..." error should not be present, as your app should have access to the `global.__UNISTYLES__` object.

Your project must be failing during the pod install or autolinking process. Ensure that you can see the success message while installing Unistyles.

**Reply by mkrotel14** [reply]
I see, for the moment I am trying to integrate into Android, sorry to not make this clear at first.

I can see that when I run a gradle sync my project has these libraries linked to it:

![Captura de Tela 2024-04-26 às 10 05 28](https://github.com/jpudysz/react-native-unistyles/assets/28939811/100eeaed-08df-4293-b6ab-c20ee8377a1b)

They are indeed installed and present in the package.json and the auto-link process ran without issues, but I don't see the react-native-unistyle... it is true that some libraries that I also have installed and present in the package.json are not showing as well in the Android project but still works, for instance, react-intl.

There is also the react-native-vector-icons library that I need to add an apply to my project build.gradle and I tried to do the same with Unistyles but to no avail.

```build.gradle
apply from: file("$rootDir/../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesAppBuildGradle(project)
apply from: file("$rootDir/../node_modules/react-native-vector-icons/fonts.gradle")
//apply from: file("$rootDir/../node_modules/react-native-unistyles/android/build.gradle")
```
The path looks weird because Android/iOS projects are a submodule of my react-native project, so they don't share the same repository, I just copy `node_modules`, `package.json` and `react-native.config.js` when building on the Android project.

Do you know if this library allows for manual linking?

My Main application has no other Packages (other than the ones I wrote for specific features)
```kt
    private fun getCustomPackages(): MutableList<ReactPackage> {
        return PackageList(this).packages.toMutableList()
    }

    private val reactNativeHost: DefaultReactNativeHost = object : DefaultReactNativeHost(this) {
        override fun getJSMainModuleName() = "index"
        override fun getBundleAssetName() = "index.android.bundle"
        override fun getUseDeveloperSupport() = BuildConfig.DEBUG
        override fun getJavaScriptExecutorFactory() = HermesExecutorFactory()
        override fun getPackages() = getCustomPackages()
        override val isHermesEnabled: Boolean = BuildConfig.IS_HERMES_ENABLED
        override val isNewArchEnabled: Boolean = BuildConfig.IS_NEW_ARCHITECTURE_ENABLED
    }
```

Edit: forgot to mention but the erro when the import unistyle is present is 
```bash
 ERROR  Error: Unistyles runtime is not available. Make sure you followed the installation instructions, js engine: hermes
 LOG  Running "Feature1" with {"rootTag":11}
 ERROR  Invariant Violation: "Feature1" has not been registered. This can happen if:
* Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
* A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called., js engine: hermes
```
But when I comment the import the app renders and no error is present in the metro bundler

**Reply by jpudysz** [reply]
It allows for manual linking. I'm using manual linking in react-native-unistyles repository (for local example). You can use react-native.config.js file and specify the uri to the node modules package. Here is example: https://github.com/jpudysz/react-native-unistyles/blob/main/examples/expo/react-native.config.js

**Reply by mkrotel14** [reply]
I ran some tests and I can assert, as you mentioned, that the issue is with autolink... the weird thing is that other installed libraries are being autolinked without issues and I don't need to manually import this module.

I was able to import this module on `File -> New -> Import Module` or the good old:
```gradle
// settings.gradle

include ':react-native-unistyles'

project(':react-native-unistyles').projectDir = new File(rootDir, '../node_modules/react-native-unistyles/android')
```
Next step is to test on iOS but if there is a `podspec` file, probably will be easier.

As to why autolink didn't work, still a mistery, but at least I can use the library and not rely on styled-components anymore 🙏

**Reply by jpudysz** [reply]
Cool! You can also play around with react-native.config.js. It should be the easiest way to link it

**Reply by mkrotel14** [reply, marked as answer]
Nice, it worked 🎉 

Although a little bit different from the example you provided, I looked on what properties and how to use it for `react-native.config.js` file. This is how it looks like:

```js
module.exports = {
  dependency: {
    platforms: {
      android: {
        sourceDir: './node_modules/react-native-unistyles/android'
      }
    }
  }
}
````
After that, i just did a Gradle Sync on Android Studio and voilà.
